src/libcore/tests/iter.rs

```diff
@@ -2144,3 +2144,16 @@ fn test_monad_laws_associativity() {
     assert_eq!((0..10).flat_map(f).flat_map(g).sum::<usize>(),
                 (0..10).flat_map(|x| f(x).flat_map(g)).sum::<usize>());
 }
+
+#[test]
+fn test_is_sorted() {
+    assert!([1, 2, 2, 9].iter().is_sorted());
+    assert!(![1, 3, 2].iter().is_sorted());
+    assert!([0].iter().is_sorted());
+    assert!(std::iter::empty::<i32>().is_sorted());
+    assert!(![0.0, 1.0, std::f32::NAN].iter().is_sorted());
+    assert!([-2, -1, 0, 3].iter().is_sorted());
+    assert!(![-2i32, -1, 0, 3].iter().is_sorted_by_key(|n| n.abs()));
+    assert!(!["c", "bb", "aaa"].iter().is_sorted());
+    assert!(["c", "bb", "aaa"].iter().is_sorted_by_key(|s| s.len()));
+}
```

Comment:
Thanks for pointing this out; I was not aware that `.iter()` was returning an `Iter<T>` instead of an `Iterator<T>`. Is changing these to `Vec` (e.g. `vec!["c", "bb", "aaa"].iter()`) enough to make sure it dispatches to the implementations of the `Iterator` trait?

Reply:
The `Iterator` impl for `slice::Iter` is generated with [this macro](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L2768). The macro is invoked for `slice::Iter` [here](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L3083) and for `slice::IterMut` [here](https://github.com/rust-lang/rust/blob/52cc6fde4b73a3013ce2185052f9492ae3728367/src/libcore/slice/mod.rs#L3184). I think we can override `is_sorted` in both impls (`Iter` and `IterMut`). So the method can just be added in the `iterator` macro.